### PR TITLE
[Bug 20239] Ensure answer folder shows the prompt on OSX 10.11+

### PR DESCRIPTION
**Pull request by livecodepanos**

### CODE REVIEW

**Comment by livecodepanos** [comment]
Does this have to be rebased into `develop-8.2`?

**Comment by runrevmark** [comment]
I suggest fixing things against develop-8.1, until 8.2 gets to RC - we can merge forward changes more easily than back.

File: engine/src/mac-dialog.mm [hunk 1]
```diff
@@ -219,14 +219,28 @@ void MCPlatformBeginFolderDialog(MCPlatformWindowRef p_owner, MCStringRef p_titl
...
-        [t_panel setTitle: [NSString stringWithMCStringRef: p_prompt]];
+    {
+        extern uint4 MCmajorosversion;
+        if (MCmajorosversion >= 0x10B0)
+            [t_panel setMessage: [NSString stringWithMCStringRef: p_prompt]];
+        else
+            [t_panel setTitle: [NSString stringWithMCStringRef: p_prompt]];
+        
+    }
+        //[t_panel setTitle: [NSString stringWithMCStringRef: p_prompt]];
```
Comment (runrevmark) [inline engine/src/mac-dialog.mm]: You can remove the commented out line.

Also, I suggest we add a comment before the initial 'if':

    If we have both a title and a message, then 'title' is used on setTitle, and 'prompt' is used for message.

    If there is no title, then on 10.11+ we must use the message field of the dialog; on <10.11 we must use the title field

Also, what happens in other combinations on <10.11 and >=10.11? i.e.

No title, no prompt.

No title, prompt.

Title, no prompt.

Title, Prompt.

(Just so we can ensure all bases are covered).

Comment (livecodepanos) [reply engine/src/mac-dialog.mm]: Oops I missed removing the commented-out line.

With this patch, the behavior  is as follows:

OSX < 10.11, all those cases behave as expected. If no title and no prompt are present, you only see an "Open" message in the top of the dialog

OSX >= 10.11:

No title, no promp -> nothing is shown
No title, prompt -> prompt is shown
Title, no prompt -> nothing is shown  (--> this probably needs tweaking?)
Title, Prompt -> prompt is shown

So in other words, on OSX >=10.11 with this patch (as with patch https://github.com/livecode/livecode/pull/3789 for the `answer file` case), only the prompt is shown, and the title is ignored.

Comment (runrevmark) [reply engine/src/mac-dialog.mm]: Yes - so I think the behavior should be that 'prompt' takes precedence over 'title'. So we ensure at least the 'prompt' is shown in all cases. Could we abstract the routine here so it can be used in all cases an NSOpenPanel is used? That way we only have to maintain one place where these decisions have to be made.

Comment (runrevmark) [reply engine/src/mac-dialog.mm]: Also, we should update the docs to explicitly explain this situation - and make sure it is clear that you should use 'prompt' for any critical information, as 'title' may or may not appear, depending on platform.

File: engine/src/mac-dialog.mm [hunk 2]
```diff
@@ -605,70 +563,82 @@ - (BOOL)panel:(id)sender shouldShowFilename:(NSString *)filename
 
 ////////////////////////////////////////////////////////////////////////////////
 
-void MCPlatformBeginFileDialog(MCPlatformFileDialogKind p_kind, MCPlatformWindowRef p_owner, MCStringRef p_title, MCStringRef p_prompt, MCStringRef *p_types, uint4 p_type_count, MCStringRef p_initial)
+void MCPlatformBeginFolderOrFileDialog(MCPlatformFileDialogKind p_kind, MCPlatformWindowRef p_owner, MCStringRef p_title, MCStringRef p_prompt, MCStringRef *p_types, uint4 p_type_count, MCStringRef p_initial, bool p_is_file)
```
Comment (runrevmark) [inline engine/src/mac-dialog.mm]: Does this need the 'p_is_file' parameter, since the p_kind parameter can already indicate a folder dialog?

File: engine/src/desktop-ans.cpp [hunk 3]
```diff
@@ -60,7 +60,7 @@ int MCA_folder(MCStringRef p_title, MCStringRef p_prompt, MCStringRef p_initial,
 	MCPlatformWindowRef t_owner;
     t_owner = compute_sheet_owner(p_options);
 	
-	MCPlatformBeginFolderOrFileDialog(kMCPlatformFileDialogKindFolder, t_owner, p_title, p_prompt, nil, 1, p_initial, false);
+	MCPlatformBeginFolderOrFileDialog(kMCPlatformFileDialogKindFolder, t_owner, p_title, p_prompt, nil, 1, p_initial);
```
Comment (runrevmark) [inline engine/src/desktop-ans.cpp]: If the p_types parameter is nullptr, shouldn't the type_count be 0? (I suspect there's a nullptr guard in the function, but should probably make it correct in the caller here).

File: engine/src/mac-dialog.mm [hunk 4]
```diff
@@ -563,7 +563,7 @@ - (BOOL)panel:(id)sender shouldShowFilename:(NSString *)filename
 
 ////////////////////////////////////////////////////////////////////////////////
 
-void MCPlatformBeginFolderOrFileDialog(MCPlatformFileDialogKind p_kind, MCPlatformWindowRef p_owner, MCStringRef p_title, MCStringRef p_prompt, MCStringRef *p_types, uint4 p_type_count, MCStringRef p_initial, bool p_is_file)
+void MCPlatformBeginFolderOrFileDialog(MCPlatformFileDialogKind p_kind, MCPlatformWindowRef p_owner, MCStringRef p_title, MCStringRef p_prompt, MCStringRef *p_types, uint4 p_type_count, MCStringRef p_initial)
```
Comment (runrevmark) [inline engine/src/mac-dialog.mm]: For clarity, perhaps the order of params should be:

kind, owner, title, prompt, initial, types = nullptr, type_count = 0

Using default params, will make the use of this in the case of being a folder dialog easier to read.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok c9f2f9e

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

**Comment by livecode-vulcan** [comment]
😞 test failure c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

* try-community-armv6-android-api8: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/571)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/585)
* try-community-js-emscripten-sdk1.35: [failure](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/559)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/502)
* try-community-universal-ios-iphonesimulator10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/505)
* try-community-universal-mac-macosx10.6: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/565)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/583)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/570)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/585)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/569)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/578)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/600)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/564)

**Comment by livecodepanos** [comment]
@livecode-vulcan retry c9f2f9e

**Comment by livecode-vulcan** [comment]
😞 test failure c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/574)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/588)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/562)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/505)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/508)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/568)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/586)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/573)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/588)
* try-community-x86_64-linux-debian7: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/572)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/581)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/603)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/567)

**Comment by livecodepanos** [comment]
@livecode-vulcan retry c9f2f9e

**Comment by livecode-vulcan** [comment]
😞 test failure c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

* try-community-armv6-android-api8: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/579)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/593)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/567)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/510)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/513)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/573)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/591)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/578)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/593)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/577)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/586)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/608)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/572)

**Comment by runrevmark** [comment]
@livecode-vulcan retry c9f2f9e

**Comment by livecode-vulcan** [comment]
😞 test failure c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/583)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/597)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/571)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/514)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/517)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/577)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/595)
* try-community-x86-linux-debian7: [failure](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/582)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/597)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/581)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/590)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/612)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/576)

**Comment by livecodepanos** [comment]
@livecode-vulcan retry c9f2f9e

**Comment by livecode-vulcan** [comment]
😞 test failure c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/585)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/599)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/573)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/516)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/519)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/579)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/597)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/584)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/599)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/583)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/592)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/614)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/578)

**Comment by livecodepanos** [comment]
@livecode-vulcan retry c9f2f9e

**Comment by livecode-vulcan** [comment]
😎 test success c9f2f9e0e317526a7a7f1fa14b7a73a6db772e43

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/586)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/600)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/574)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/517)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/520)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/580)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/598)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/585)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/600)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/584)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/593)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/615)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/579)
